packages/opentelemetry-plugin-xml-http-request/src/xhr.ts

```diff
@@ -0,0 +1,370 @@
+/*!
+ * Copyright 2019, OpenTelemetry Authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import {
+  B3Format,
+  BasePlugin,
+  isUrlIgnored,
+  isWrapped,
+  otperformance,
+  urlMatches,
+} from '@opentelemetry/core';
+import * as tracing from '@opentelemetry/tracing';
+import * as types from '@opentelemetry/types';
+import {
+  addSpanNetworkEvent,
+  getResource,
+  parseUrl,
+  PerformanceTimingNames as PTN,
+} from '@opentelemetry/web';
+import * as shimmer from 'shimmer';
+import { AttributeNames } from './enums/AttributeNames';
+import { EventNames } from './enums/EventNames';
+import {
+  OpenFunction,
+  PropagateTraceHeaderUrls,
+  SendBody,
+  SendFunction,
+  XMLHttpRequestWrapped,
+} from './types';
+
+/**
+ * XMLHttpRequest config
+ */
+export interface XMLHttpRequestPluginConfig extends types.PluginConfig {
+  // urls which should include trace headers when origin doesn't match
+  propagateTraceHeaderUrls?: PropagateTraceHeaderUrls;
+}
+
+/**
+ * This class represents a XMLHttpRequest plugin for auto instrumentation
+ */
+export class XMLHttpRequestPlugin extends BasePlugin<XMLHttpRequest> {
+  readonly component: string = 'xml-http-request';
+  readonly version: string = '0.3.0';
+  moduleName = this.component;
+
+  protected _config!: XMLHttpRequestPluginConfig;
+
+  private _callbackToRemoveEvents: { [key: string]: Function } = {};
+  private _spans: { [key: string]: tracing.Span } = {};
+  private _usedResources: { [key: string]: PerformanceResourceTiming[] } = {};
```

Comment:
I think there is a similar memory leak risk here. Entries are added in `_markResourcesAsUsed` but are never deleted

Reply:
these are all cleaned now whenever it can

Reply:
converted into WeakSet